### PR TITLE
Run MC cmds via terminal & serv.commands

### PR DESCRIPTION
**Pull request by redcarti**
Pull request of #401

### CODE REVIEW

**Comment by redcarti** [comment]
It needs to be in other files but whatever..

**Comment by redcarti** [comment]
And it needs to be rewrited... cuz 
image

**Comment by redcarti** [comment]
hmm, thats strange

**Comment by redcarti** [comment]
@rom1504 are throwing an error is needed? I think that just sending message is better.
Also, about ``server.selectorString``, why do it need ``pos``? E.g. ``/kill @a`` don't needs the ``pos``.

**Comment by redcarti** [comment]
Oh, I see that ``handleCommand`` checks ``UserError``. Whatever, will change.

**Comment by redcarti** [comment]
Needed some more cmds..

**Comment by redcarti** [comment]
Input needs to be at the end of console. But that's a problem...

**Comment by redcarti** [comment]
Alright, I tried many different ways to make ``process.stdin`` cursor to the end of a console. But I didn't find best way to do this.

**Comment by redcarti** [comment]
I think that's all for this feature. It can be upgraded.

**Comment by redcarti** [comment]
@rom1504 so?

**Comment by redcarti** [comment]
I think we could add to cmds ``player.chat`` thing, e.g. when executed ``/op <player>``, player that opped will recieve _[<opplayer>: Opped <player>]_ or sth. It's implemented in my ``input.js`` plugin.

**Comment by rom1504** [comment]
Looks good, I'll review it when I have some time

**Comment by redcarti** [comment]
Okay, I'll wait. About last comment, I'll make it soon and create pull request.

**Comment by rom1504** [comment]
Ok reviewed.
Overall it looks good but there is one point that I think would improve this whole feature.

Instead of duplicating all the commands between player and server, what about defining them once and putting them both on the player and on the server ?
If you look at https://github.com/PrismarineJS/flying-squid/pull/402/files#diff-3d81549536b6e81c8b52be79354cdf85R236 vs https://github.com/PrismarineJS/flying-squid/blob/43b665bb84afc44f58758671d5a1e8bc75809cbe/src/lib/plugins/moderation.js#L188 for example, it's exactly the same thing.

The only difference is that in the second case, player is coming from the context, whereas in the first case player is found from the argument.
It should be possible to define it all once and use it to do serv.commands.add and player.commands.add (or even better, make a single .add that will do both)

Do you think you could try doing this ?

I think this is the right time to do this, before everything is duplicated and hard to merge

**Comment by redcarti** [comment]
I think, that might be possible, adding the feature of context in ``Command`` class. Or you meant something different?

**Comment by redcarti** [comment]
I think ``serv.commands`` instead of ``player.commands`` might be better.

**Comment by redcarti** [comment]
And, ``player.handleCommand`` will be using ``serv.commands.use()``

**Comment by redcarti** [comment]
And yeah, so I'm in PrismarineJS discord, so discussion about this feature might be faster

**Comment by redcarti** [comment]
So,
```js
player.handleCommand = async (str) => {
    try {
      const res = await serv.commands.use(str, player)
      if (res) player.chat(serv.color.red + res)
    } catch (err) {
      if (err.userError) player.chat(serv.color.red + 'Error: ' + err.message)
      else setTimeout(() => { throw err }, 0)
    }
  }

serv.handleCommand = async (str) => {
    try {
      const res = await serv.commands.use(str)
      if (res) serv.log('[INFO]: ' + res)
    } catch (err) {
      if (err.userError) serv.log('[ERR]: ' + err.message)
      else setTimeout(() => { throw err }, 0)
    }
  }
```

File: src/lib/plugins/commands.js [hunk 1]
```diff
@@ -1,81 +1,105 @@
...
+  serv.handleCommand = async (str) => {
+    try {
+      const res = await serv.commands.use(str)
+      if (res) serv.info(res)
+    } catch (err) {
+      if (err.userError) serv.err(err.message)
+      else setTimeout(() => { throw err }, 0)
+    }
+  }
+  /*
```
Comment (rom1504) [inline src/lib/plugins/commands.js:30]: Why is this commented ?

File: src/lib/plugins/pvp.js [hunk 2]
```diff
@@ -30,7 +30,8 @@ module.exports.player = function (player, serv) {
     if (mouse === 1) { attackEntity(target) }
   })
 
-  player.commands.add({
+  /*
```
Comment (rom1504) [inline src/lib/plugins/pvp.js:33]: Why commented ?

File: src/lib/plugins/blocks.js [hunk 3]
```diff
@@ -60,14 +62,14 @@ module.exports.player = function (player, serv) {
       if (!results) return false
       return results
     },
-    action (params) {
+    action (params, ctx) {
       let res = params.slice(1, 4)
-      res = res.map((val, i) => serv.posFromString(val, player.position[['x', 'y', 'z'][i]]))
-      player.setBlock(new Vec3(res[0], res[1], res[2]).floored(), params[4], params[5] || 0)
+      res = res.map((val, i) => serv.posFromString(val, ctx.player.position[['x', 'y', 'z'][i]]))
```
Comment (rom1504) [inline src/lib/plugins/blocks.js:67]: You need to either do something appropriate if ctx.player is undefined or mark this as playerOnly

Same for all the commands

File: src/lib/plugins/commands.js (hunk 1 above)
Comment (redcarti) [reply src/lib/plugins/commands.js:30]: forgot to delete this

File: src/lib/plugins/pvp.js (hunk 2 above)
Comment (redcarti) [reply src/lib/plugins/pvp.js:33]: as gm, forgot to delete

File: src/lib/plugins/blocks.js (hunk 3 above)
Comment (redcarti) [reply src/lib/plugins/blocks.js:67]: will work on it soon

File: src/lib/plugins/blocks.js [hunk 4]
```diff
@@ -60,14 +62,16 @@ module.exports.player = function (player, serv) {
       if (!results) return false
       return results
     },
-    action (params) {
+    action (params, ctx) {
       let res = params.slice(1, 4)
-      res = res.map((val, i) => serv.posFromString(val, player.position[['x', 'y', 'z'][i]]))
-      player.setBlock(new Vec3(res[0], res[1], res[2]).floored(), params[4], params[5] || 0)
+      if (ctx.player) res = res.map((val, i) => serv.posFromString(val, ctx.player.position[['x', 'y', 'z'][i]]))
+      else res = res.map((val, i) => serv.posFromString(val, new Vec3(0, 128, 0)[['x', 'y', 'z'][i]]))
```
Comment (rom1504) [inline src/lib/plugins/blocks.js:68]: Why this position ? Does the vanilla server do that ?

File: src/lib/plugins/commands.js [hunk 5]
```diff
@@ -1,81 +1,59 @@
...
-  player.commands.add({
+  serv.commands.add({
     base: 'modpe',
     info: 'for modpe commands',
     usage: '/modpe <params>',
     parse (str) { return str || false },
-    action (str) {
-      player.emit('modpe', str)
+    action (str, ctx) {
+      ctx.player.emit('modpe', str)
```
Comment (rom1504) [inline src/lib/plugins/commands.js:52]: Handle ctx.player undefined or mark as player only

File: src/lib/plugins/commands.js [hunk 6]
```diff
@@ -93,36 +71,102 @@ module.exports.player = function (player, serv, { version }) {
...
     usage: '/selector <selector>',
     op: true,
     parse (str) {
       return str || false
     },
-    action (sel) {
-      const arr = serv.selectorString(sel, player.position, player.world)
-      player.chat(JSON.stringify(arr.map(a => a.id)))
+    action (sel, ctx) {
+      const arr = serv.selectorString(sel, ctx.player.position, ctx.player.world)
```
Comment (rom1504) [inline src/lib/plugins/commands.js:83]: Handle ctx.player undefined

File: src/lib/plugins/daycycle.js [hunk 7]
```diff
@@ -48,9 +46,9 @@ module.exports.player = function (player, serv) {
         value: data[2] === 'day' ? 1000 : (data[2] === 'night' ? 13000 : parseInt(data[2]))
       }
     },
-    action ({ action, value }) {
+    action ({ action, value }, ctx) {
       if (action === 'query') {
-        player.chat('It is ' + serv.time)
+        ctx.player.chat('It is ' + serv.time)
```
Comment (rom1504) [inline src/lib/plugins/daycycle.js:51]: Handle ctx.player undefined

File: src/lib/plugins/effects.js [hunk 8]
```diff
@@ -50,16 +50,16 @@ module.exports.entity = function (entity, serv) {
...
     base: 'effect',
     info: 'Give player an effect',
     usage: '/effect <player> <effect> [seconds] [amplifier] [hideParticles]',
     parse (str) {
       return str.match(/(.+?) (\d+)(?: (\d+))?(?: (\d+))?(?: (true|false))?|.*? clear/) || false
     },
-    action (params) {
-      const targets = player.selectorString(params[1])
+    action (params, ctx) {
+      const targets = ctx.player.selectorString(params[1])
```
Comment (rom1504) [inline src/lib/plugins/effects.js:62]: Handle ctx.player undefined

File: src/lib/plugins/effects.js [hunk 9]
```diff
@@ -78,9 +78,9 @@ module.exports.player = function (player) {
         })
       }
       const chatSelect = (targets.length === 1 ? (targets[0].type === 'player' ? targets[0].username : 'entity') : 'entities')
-      if (params[2] === 'clear') player.chat('Remove all effects from ' + chatSelect + '.')
+      if (params[2] === 'clear') ctx.player.chat('Remove all effects from ' + chatSelect + '.')
```
Comment (rom1504) [inline src/lib/plugins/effects.js:81]: Handle ctx.player undefined

File: src/lib/plugins/experience.js [hunk 10]
```diff
@@ -29,30 +29,32 @@ module.exports.player = function (player, serv) {
...
       const amt = parseInt(args[1])
-      const user = args[3] ? serv.getPlayer(args[3]) : player
+      const user = args[3] ? serv.getPlayer(args[3]) : ctx.player
       if (!user) return args[3] + ' is not on this server!'
 
       if (!isLevel) {
         user.setXp(user.xp + amt)
-        player.chat('Gave ' + user.username + ' ' + amt + ' xp')
+        ctx.player.chat('Gave ' + user.username + ' ' + amt + ' xp')
       } else {
```
Comment (rom1504) [inline src/lib/plugins/experience.js:52]: Handle ctx.player undefined

**Comment by redcarti** [comment]
idk, if I type ``setblock ~1 ~1 ~1 stone``, server will return ``Cannot place block outside of the world``

File: src/lib/plugins/moderation.js [hunk 11]
```diff
@@ -89,18 +137,18 @@ module.exports.player = function (player, serv) {
         reason: parts.join(' ')
       }
     },
-    action ({ username, reason }) {
+    action ({ username, reason }, ctx) {
       const kickPlayer = serv.getPlayer(username)
       if (!kickPlayer) {
-        player.chat(username + ' is not on this server!')
+        ctx.player.chat(username + ' is not on this server!')
```
Comment (rom1504) [inline src/lib/plugins/moderation.js:143]: Handle ctx.player undefined

File: src/lib/plugins/moderation.js [hunk 12]
```diff
@@ -113,28 +161,28 @@ module.exports.player = function (player, serv) {
...
+    action ({ username, reason }, ctx) {
       const banPlayer = serv.getPlayer(username)
 
       if (!banPlayer) {
         serv.banUsername(username, reason)
           .then(() => {
-            serv.emit('banned', player, username, reason)
-            player.chat(username + ' was banned')
+            serv.emit('banned', ctx.player, username, reason)
+            ctx.player.chat(username + ' was banned')
```
Comment (rom1504) [inline src/lib/plugins/moderation.js:171]: Handle ctx.player undefined

File: src/lib/plugins/moderation.js [hunk 13]
```diff
@@ -148,24 +196,24 @@ module.exports.player = function (player, serv) {
...
+  serv.commands.add({
     base: 'pardon-ip',
     info: 'to pardon a player by ip',
     usage: '/pardon-ip <ip>',
     op: true,
-    action (IP) {
+    action (IP, ctx) {
       const result = serv.pardonIP(IP)
-      player.chat(result ? IP + ' was IP pardoned' : IP + ' is not banned')
+      ctx.player.chat(result ? IP + ' was IP pardoned' : IP + ' is not banned')
```
Comment (rom1504) [inline src/lib/plugins/moderation.js:212]: Handle ctx.player undefined

**Comment by redcarti** [comment]
So, for a while it might be ``0, 128, 0``

**Comment by redcarti** [comment]
Okay, workin on it

File: src/lib/plugins/blocks.js (hunk 4 above)
Comment (redcarti) [reply src/lib/plugins/blocks.js:68]: idk, if I type ``setblock ~1 ~1 ~1 stone``, server will return ``Cannot place block outside of the world``

**Comment by redcarti** [comment]
wait

File: src/lib/plugins/commands.js (hunk 5 above)
Comment (redcarti) [reply src/lib/plugins/commands.js:53]: ``onlyPlayer: true``

File: src/lib/plugins/commands.js [hunk 14]
```diff
@@ -93,36 +72,103 @@ module.exports.player = function (player, serv, { version }) {
...
 
-  player.handleCommand = async (str) => {
-    try {
-      const res = await player.commands.use(str, player.op)
-      if (res) player.chat(serv.color.red + res)
-    } catch (err) {
-      if (err.userError) player.chat(serv.color.red + 'Error: ' + err.message)
-      else setTimeout(() => { throw err }, 0)
+  serv.commands.add({
+    base: 'help',
```
Comment (rom1504) [inline src/lib/plugins/commands.js:91]: Could you exclude only player from server help and only console from player help ?

File: src/lib/plugins/experience.js [hunk 15]
```diff
@@ -29,30 +29,34 @@ module.exports.player = function (player, serv) {
...
         user.setXp(user.xp + amt)
-        player.chat('Gave ' + user.username + ' ' + amt + ' xp')
+        if(ctx.player) ctx.player.chat('Gave ' + user.username + ' ' + amt + ' xp')
+        else serv.log('Gave ' + user.username + ' ' + amt + ' xp')
       } else {
-        const currLevel = getXpLevel(player.xp)
+        const currLevel = getXpLevel(ctx.player.xp)
         const baseCurrLevel = getBaseXpFromLevel(currLevel)
-        const extraXp = player.xp - baseCurrLevel
+        const extraXp = ctx.player.xp - baseCurrLevel
```
Comment (rom1504) [inline src/lib/plugins/experience.js:56]: Check this one too

File: src/lib/plugins/moderation.js [hunk 16]
```diff
@@ -89,18 +137,19 @@ module.exports.player = function (player, serv) {
...
+    action ({ username, reason }, ctx) {
       const kickPlayer = serv.getPlayer(username)
       if (!kickPlayer) {
-        player.chat(username + ' is not on this server!')
+        if(ctx.player) ctx.player.chat(username + ' is not on this server!')
+        else throw new UserError(username + ' is not on this server!')
       } else {
         kickPlayer.kick(reason)
-        kickPlayer.emit('kicked', player, reason)
+        kickPlayer.emit('kicked', ctx.player, reason)
```
Comment (rom1504) [inline src/lib/plugins/moderation.js:147]: Need to check this one

File: src/lib/plugins/moderation.js [hunk 17]
```diff
@@ -148,24 +199,25 @@ module.exports.player = function (player, serv) {
         reason: argv.shift()
       }
     },
-    action ({ IP, reason }) {
+    action ({ IP, reason }, ctx) {
       serv.banIP(IP, reason)
-      player.chat('' + IP + ' was IP banned')
+      ctx.player.chat('' + IP + ' was IP banned')
```
Comment (rom1504) [inline src/lib/plugins/moderation.js:204]: Handle undefined ctx player

File: src/lib/plugins/moderation.js [hunk 18]
```diff
@@ -174,48 +226,33 @@ module.exports.player = function (player, serv) {
       if (!str.match(/([a-zA-Z0-9_]+)/)) { return false }
       return str
     },
-    action (nick) {
+    action (nick, ctx) {
       serv.pardonUsername(nick)
-        .then(() => player.chat(nick + ' is unbanned'))
+        .then(() => ctx.player.chat(nick + ' is unbanned'))
```
Comment (rom1504) [inline src/lib/plugins/moderation.js:231]: Handle this one

File: src/lib/plugins/particle.js [hunk 19]
```diff
@@ -39,13 +37,13 @@ module.exports.player = function (player, serv) {
         size: results[5] ? new Vec3(parseInt(results[3]), parseInt(results[4]), parseInt(results[5])) : new Vec3(1, 1, 1)
       }
     },
-    action ({ particle, amount, size }) {
+    action ({ particle, amount, size }, ctx) {
       if (amount >= 100000) {
-        player.chat('You cannot emit more than 100,000 particles!')
+        ctx.player.chat('You cannot emit more than 100,000 particles!')
         return
```
Comment (rom1504) [inline src/lib/plugins/particle.js:43]: Handle this

File: src/lib/plugins/physics.js [hunk 20]
```diff
@@ -96,17 +96,17 @@ module.exports.entity = function (entity, serv, { version }) {
...
     info: 'Push velocity on player(s)',
     usage: '/velocity <player> <x> <y> <z>',
     op: true,
     parse (str) {
       return str.match(/(.+?) (\d+) (\d+) (\d+)/) || false
     },
-    action (params) {
-      const selector = player.selectorString(params[1])
+    action (params, ctx) {
+      const selector = ctx.player.selectorString(params[1])
```
Comment (rom1504) [inline src/lib/plugins/physics.js:109]: Handle this

File: src/lib/plugins/experience.js (hunk 15 above)
Comment (redcarti) [reply src/lib/plugins/experience.js:56]: there might be user, not player.

File: src/lib/plugins/commands.js (hunk 14 above)
Comment (redcarti) [reply src/lib/plugins/commands.js:91]: I think it might be ok

File: src/lib/plugins/moderation.js (hunk 16 above)
Comment (redcarti) [reply src/lib/plugins/moderation.js]: fixed

File: src/lib/plugins/moderation.js (hunk 17 above)
Comment (redcarti) [reply src/lib/plugins/moderation.js]: fixed

File: src/lib/plugins/moderation.js (hunk 18 above)
Comment (redcarti) [reply src/lib/plugins/moderation.js]: fixed

File: src/lib/plugins/particle.js (hunk 19 above)
Comment (redcarti) [reply src/lib/plugins/particle.js:44]: only player

File: src/lib/plugins/commands.js (hunk 14 above)
Comment (redcarti) [reply src/lib/plugins/commands.js:91]: excluding will reduce page length... so added player&console only

**Pull request merged**